File: timely-bench/src/bin/stream_bench.rs
//! Timely Dataflow STREAM benchmark.
//!
//! Each Timely worker independently allocates and operates on its own copy of
//! the arrays (total memory = workers * array_size * bytes_per_element).
//! This matches the SynStream STREAM setup: each worker runs one instance of
//! the kernel on an independent array.
//!
//! Two variants:
//!   (default) fresh allocation per rep  — system label "timely"
//!   --pooled  pre-allocated, reused     — system label "timely_pooled"
//!
//! Usage:
//!   cargo run -r -p timely-bench --bin stream_bench -- \
//!     --kernel triad --array-size 268435456 --workers 4 --output timely_stream.csv
//!   cargo run -r -p timely-bench --bin stream_bench -- \
//!     --kernel triad --array-size 268435456 --workers 4 --pooled --output timely_pooled_stream.csv

use clap::Parser;
use std::sync::{Arc, Barrier, Mutex};
use std::time::Instant;
use timely_bench::append_csv;

#[derive(Parser, Debug)]
#[command(name = "stream_bench", about = "Timely Dataflow STREAM benchmark")]
struct Cli {
    /// STREAM kernel: copy | scale | add | triad
    #[arg(long, value_parser = ["copy", "scale", "add", "triad"], default_value = "triad")]
    kernel: String,

    /// Array size per worker (number of f64 elements)
    #[arg(long, default_value_t = 268_435_456usize)]
    array_size: usize,

    /// Number of Timely worker threads
    #[arg(long, default_value_t = 1usize)]
    workers: usize,

    /// Scalar for scale/triad kernels
    #[arg(long, default_value_t = 3.0f64)]
    scalar: f64,

    /// Output CSV file
    #[arg(long, default_value = "timely_stream.csv")]
    output: String,

    /// Measurement repetitions (warm-up excluded from stats)
    #[arg(long, default_value_t = 20usize)]
    reps: usize,

    /// Warm-up repetitions
    #[arg(long, default_value_t = 3usize)]
    warmup: usize,

    /// Pre-allocate arrays once and reuse across reps (pooled variant).
    /// Eliminates per-rep malloc/free overhead; system label becomes "timely_pooled".
    #[arg(long, default_value_t = false)]
    pooled: bool,
}

fn main() {
    let cli = Cli::parse();
    let kernel = cli.kernel.clone();
    let array_size = cli.array_size;
    let num_workers = cli.workers;
    let scalar = cli.scalar;
    let reps = cli.reps;
    let warmup = cli.warmup;
    let output = cli.output.clone();

    let arrays = match kernel.as_str() {
        "copy" | "scale" => 2usize,
        "add" | "triad" => 3usize,
        _ => unreachable!(),
    };
    // bytes per repetition = workers * arrays * array_size * 8 bytes
    let bytes_total = num_workers * arrays * array_size * 8;

    let system_label = if cli.pooled { "timely_pooled" } else { "timely" };

    // Use a barrier so all workers start the timed section simultaneously
    let barrier = Arc::new(Barrier::new(num_workers));
    // Shared elapsed time: max across workers (wall-clock of last worker)
    let elapsed_vec: Arc<Mutex<Vec<f64>>> = Arc::new(Mutex::new(Vec::new()));

    let mut all_elapsed: Vec<f64> = Vec::with_capacity(warmup + reps);

    if cli.pooled {
        // --- Pooled variant: pre-allocate once, reuse across reps ---
        //
        // Each worker owns its slot: pool_X[worker.index()].
        // pool_a uses Mutex for borrow-checker satisfaction; the lock is
        // never actually contended (each worker accesses a distinct index).
        let pool_b: Arc<Vec<Vec<f64>>> = Arc::new(
            (0..num_workers).map(|_| vec![2.0f64; array_size]).collect(),
        );
        // pool_c only needed for add/triad; None for copy/scale.
        let pool_c: Arc<Option<Vec<Vec<f64>>>> = Arc::new(
            if matches!(kernel.as_str(), "add" | "triad") {
                Some((0..num_workers).map(|_| vec![1.0f64; array_size]).collect())
            } else {
                None
            },
        );
        let pool_a: Arc<Vec<Mutex<Vec<f64>>>> = Arc::new(
            (0..num_workers).map(|_| Mutex::new(vec![0.0f64; array_size])).collect(),
        );

        for _rep in 0..(warmup + reps) {
            let barrier_rep = barrier.clone();
            let elapsed_rep = elapsed_vec.clone();
            let kernel_rep = kernel.clone();
            let pa = pool_a.clone();
            let pb = pool_b.clone();
            let pc = pool_c.clone();

            timely::execute_from_args(
                vec![String::from("-w"), num_workers.to_string()].into_iter(),
                move |worker| {
                    let wi = worker.index();
                    // Acquire worker-exclusive output slot (never contended).
                    let mut a = pa[wi].lock().unwrap();
                    let b = &pb[wi];

                    barrier_rep.wait();
                    let t0 = Instant::now();

                    match kernel_rep.as_str() {
                        "copy" => {
                            let n = b.len();
                            for i in 0..n { a[i] = b[i]; }
                        }
                        "scale" => {
                            let n = b.len();
                            for i in 0..n { a[i] = scalar * b[i]; }
                        }
                        "add" => {
                            let c = &pc.as_ref().as_ref().unwrap()[wi];
                            let n = b.len();
                            for i in 0..n { a[i] = b[i] + c[i]; }
                        }
                        "triad" => {
                            let c = &pc.as_ref().as_ref().unwrap()[wi];
                            let n = b.len();
                            for i in 0..n { a[i] = b[i] + scalar * c[i]; }
                        }
                        _ => unreachable!(),
                    }

                    std::hint::black_box(&*a);

                    let elapsed = t0.elapsed().as_secs_f64();
                    let mut guard = elapsed_rep.lock().unwrap();
                    guard.push(elapsed);
                },
            )
            .expect("Timely execute failed");

            let max_elapsed = {
                let mut guard = elapsed_vec.lock().unwrap();
                let v = guard.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
                guard.clear();
                v
            };
            all_elapsed.push(max_elapsed);
        }
    } else {
        // --- Base variant: fresh allocation per rep ---
        for _rep in 0..(warmup + reps) {
            let barrier_rep = barrier.clone();
            let elapsed_rep = elapsed_vec.clone();
            let kernel_rep = kernel.clone();

            timely::execute_from_args(
                vec![String::from("-w"), num_workers.to_string()].into_iter(),
                move |worker| {
                    let _wi = worker.index();
                    let n = array_size;

                    // Allocate arrays local to each worker
                    let mut a: Vec<f64> = vec![0.0; n];
                    let b: Vec<f64> = vec![2.0; n];
                    let c: Vec<f64> = vec![1.0; n];

                    // All workers wait at barrier before starting timed section
                    barrier_rep.wait();
                    let t0 = Instant::now();

                    match kernel_rep.as_str() {
                        "copy" => {
                            for i in 0..n { a[i] = b[i]; }
                        }
                        "scale" => {
                            for i in 0..n { a[i] = scalar * b[i]; }
                        }
                        "add" => {
                            for i in 0..n { a[i] = b[i] + c[i]; }
                        }
                        "triad" => {
                            for i in 0..n { a[i] = b[i] + scalar * c[i]; }
                        }
                        _ => unreachable!(),
                    }

                    // Prevent dead-code elimination
                    std::hint::black_box(&a);

                    let elapsed = t0.elapsed().as_secs_f64();
                    let mut guard = elapsed_rep.lock().unwrap();
                    guard.push(elapsed);
                },
            )
            .expect("Timely execute failed");

            let max_elapsed = {
                let mut guard = elapsed_vec.lock().unwrap();
                let v = guard.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
                guard.clear();
                v
            };
            all_elapsed.push(max_elapsed);
        }
    }

    // Exclude warm-up
    let timed = &all_elapsed[warmup..];
    let mean_elapsed = timed.iter().sum::<f64>() / timed.len() as f64;
    let gb_s = bytes_total as f64 / mean_elapsed / 1e9;

    println!(
        "{} STREAM {} | workers={} | array_size={} | mean={:.4}s | {:.2} GB/s",
        system_label, kernel, num_workers, array_size, mean_elapsed, gb_s
    );

    append_csv(&output, system_label, &kernel, array_size, num_workers, mean_elapsed, gb_s);
}

File: tbb-bench/src/stream_bench.cpp
//! Intel TBB STREAM benchmark.
//!
//! Each TBB worker independently allocates and operates on its own array
//! (total memory = workers * array_size * bytes_per_element), matching the
//! SynStream and Timely STREAM setups exactly.
//!
//! Usage:
//!   ./stream_bench --kernel triad --array-size 268435456 --workers 4 \
//!                  --output tbb_stream.csv [--pin]

#include "common.h"

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

#include <oneapi/tbb/parallel_for.h>
#include <oneapi/tbb/task_arena.h>

// ---------------------------------------------------------------------------
// CLI (hand-rolled to avoid external deps)
// ---------------------------------------------------------------------------

struct Cli {
    std::string kernel     = "triad";
    std::size_t array_size = 268'435'456UL;
    int         workers    = 1;
    double      scalar     = 3.0;
    int         reps       = 20;
    int         warmup     = 3;
    std::string output     = "tbb_stream.csv";
    bool        pin        = false;
};

static Cli parse_args(int argc, char** argv) {
    Cli c;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        auto val = [&]() -> std::string {
            if (i + 1 >= argc) throw std::runtime_error("Missing value for " + a);
            return argv[++i];
        };
        if      (a == "--kernel")     c.kernel     = val();
        else if (a == "--array-size") c.array_size = std::stoull(val());
        else if (a == "--workers")    c.workers    = std::stoi(val());
        else if (a == "--scalar")     c.scalar     = std::stod(val());
        else if (a == "--reps")       c.reps       = std::stoi(val());
        else if (a == "--warmup")     c.warmup     = std::stoi(val());
        else if (a == "--output")     c.output     = val();
        else if (a == "--pin")        c.pin        = true;
        else throw std::runtime_error("Unknown flag: " + a);
    }
    return c;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

int main(int argc, char** argv) {
    Cli cli = parse_args(argc, argv);

    const std::size_t N  = cli.array_size;
    const int         W  = cli.workers;
    const double      sc = cli.scalar;

    // Number of arrays touched per rep (for GB/s calculation)
    std::size_t n_arrays = 0;
    if (cli.kernel == "copy" || cli.kernel == "scale") n_arrays = 2;
    else if (cli.kernel == "add" || cli.kernel == "triad") n_arrays = 3;
    else throw std::runtime_error("Unknown kernel: " + cli.kernel);

    const std::size_t bytes_total =
        static_cast<std::size_t>(W) * n_arrays * N * sizeof(double);

    // Allocate per-worker arrays
    std::vector<std::vector<double>> a(W, std::vector<double>(N, 0.0));
    std::vector<std::vector<double>> b(W, std::vector<double>(N, 2.0));
    std::vector<std::vector<double>> c(W, std::vector<double>(N, 1.0));

    tbb::task_arena arena(W);
    std::unique_ptr<PinningObserver> obs;
    if (cli.pin) obs = std::make_unique<PinningObserver>(arena, 1);

    std::vector<double> all_elapsed;
    all_elapsed.reserve(cli.warmup + cli.reps);

    const std::string& kernel = cli.kernel;

    for (int rep = 0; rep < cli.warmup + cli.reps; ++rep) {
        auto t0 = std::chrono::high_resolution_clock::now();

        arena.execute([&] {
            tbb::parallel_for(
                0, W,
                [&](int w) {
                    double* __restrict__ av = a[w].data();
                    const double* __restrict__ bv = b[w].data();
                    const double* __restrict__ cv = c[w].data();

                    if (kernel == "copy") {
                        for (std::size_t i = 0; i < N; ++i) av[i] = bv[i];
                    } else if (kernel == "scale") {
                        for (std::size_t i = 0; i < N; ++i) av[i] = sc * bv[i];
                    } else if (kernel == "add") {
                        for (std::size_t i = 0; i < N; ++i) av[i] = bv[i] + cv[i];
                    } else { // triad
                        for (std::size_t i = 0; i < N; ++i) av[i] = bv[i] + sc * cv[i];
                    }
                },
                tbb::static_partitioner{}
            );
        });

        auto t1 = std::chrono::high_resolution_clock::now();
        all_elapsed.push_back(
            std::chrono::duration<double>(t1 - t0).count());
    }

    // Prevent dead-code elimination
    volatile double sink = a[0][0];
    (void)sink;

    // Exclude warmup
    double sum = 0.0;
    for (int i = cli.warmup; i < cli.warmup + cli.reps; ++i)
        sum += all_elapsed[i];
    double mean = sum / cli.reps;
    double gb_s = static_cast<double>(bytes_total) / mean / 1e9;

    std::printf(
        "TBB STREAM %s | workers=%d | array_size=%zu | mean=%.4fs | %.2f GB/s\n",
        cli.kernel.c_str(), W, N, mean, gb_s);

    append_csv(cli.output, "tbb", cli.kernel, N, W, mean, gb_s);
    return 0;
}
